### DISCUSSION TITLE
migrating off of getStaticPaths & getStaticProps

### DISCUSSION THREAD

**Discussion by sgrund14** [vercel/next.js#54306] [Help]
### Summary

I was using `getStaticPaths` & `getStaticProps` on a route I have using the nextjs page router, that generated static pages for all `/[profile]` routes. something like: 

```
export const getStaticProps: GetStaticProps = () => {
    const user = getUser(handle);
    return {
        user,
    };
};

export const getStaticPaths: GetStaticPaths = () => {
    const users = getUsers();
    return {
        paths: users.map(user => user.handle),
        fallback: false,
    };
};

```

This was built and deployed, and so generated static pages on the vercel deployment for the users on my site.

I decided to switch to using `getServerSideProps` for this (because I realized it might be a bad idea to make static pages for every single user as the app grows), so I removed `getStaticPaths` & `getStaticProps` from the route definition. However, now I'm noticing that when I navigate directly to a route that doesn't have the static handlers anymore, it is trying and failing to load the static html for the site and throwing a 500 error.

<img width="386" alt="Screenshot 2023-08-20 at 1 28 06 PM" src="https://github.com/vercel/next.js/assets/14083701/3500c744-9745-4984-9140-7fffc6807e31">

I tried deleting the .next folder on the deployment, but it is still trying to load the static asset for that route. It seems like maybe next still thinks I have `getStaticPaths` defined, and is trying to load the static asset for a page that isn't being generated anymore. Is there a way to reset the build so that it is no longer trying to load the static assets, and just hitting getServerSideProps instead?

### Additional information

I've tried looking through similar problems to see if I could find a solution
https://github.com/vercel/next.js/discussions/18967
https://stackoverflow.com/questions/65526754/how-to-clear-nextjs-getstaticpaths-cache-unpublish-a-dynamic-route

but these seem more to do with returning a 404 on deleted content. I don't want to return a 404, as in my case the content (the user profile) still exists. I just want to switch my data fetching strategy to use SSR instead of static pregeneration, so it seems like what I want to do is delete the config that is telling next to look for static files for the `[profile]` route

**Comment by icyJoseph** [comment]
>  generated static pages on the vercel deployment for the users on my site

Does this mean that you are using output export mode? Could you also share a bit more of your next.config file? Or a link to a repository?

**Reply by sgrund14** [reply]
hey @icyJoseph, not using output export mode explicitly, just added `getStaticProps` & `getStaticPaths` to my route definition, which next used to automatically generate the the static pages during the build step

here's a snippet of my nextconfig (have a private repo)

```
const nextConfig = {
    reactStrictMode: true,
    images: {
        domains: ['res.cloudinary.com'],
        disableStaticImages: true,
        deviceSizes: [384, 640, 1080, 1200],
        imageSizes: [32, 64, 96, 256],
    },
    pageExtensions: ['page.tsx', 'page.ts'],
    compiler: {
        relay: {
            src: './',
            language: 'typescript',
            artifactDirectory: '../../packages/gqll/__generated__',
        },
    },
    webpack: (config, { dev, webpack, isServer }) => {
        config.plugins.push(
            new webpack.DefinePlugin({
                __DEV__: dev,
                __IS_SERVER__: isServer,
            })
        );

        return config;
    },
    async rewrites() {
        return [
           // mixpanel rewrites
        ];
    },
    transpilePackages: [
        // bunch of packages
    ],
};

module.exports = withPlugins([withExpo, withFonts, withImages], nextConfig);
```

**Reply by icyJoseph** [reply]
Cool cool, I kind of was hoping for export mode, or a PWA setup, but it doesn't look like you have any of those... 🤔 

Is this project hosted on Vercel?

**Reply by sgrund14** [reply]
Yea, hosted on vercel

Tried using their "delete build cache" tool with no success. Seems like there is something saved on disk that is using the old getStaticPaths definition

**Reply by icyJoseph** [reply]
Ahh, what if you create a new route, does that work? Like a hello-world kind of thing that runs GSSP too. Just as a sanity check.

**Reply by sgrund14** [reply]
hmmm interesting, no that breaks too.

context: the route that i added getStaticPaths/getStaticProps too was /[profile]

the route that i just added getServerSideProps too was /about

but it fails the same way when navigating straight too it 

![Screenshot 2023-08-21 at 9 02 47 AM](https://github.com/vercel/next.js/assets/14083701/69abe869-3c7c-4e2c-b25e-d40231bd250d)

so it's matching the /[profile] route for some reason?

**Reply by icyJoseph** [reply]
What next.js version have you got?

And you say this happen when navigating straight to it? Can you do SPA transitions (from other pages) onto it?

And, do you have middleware?

**Reply by sgrund14** [reply]
you can't do SPA transitions to it 😱  it gives the same 500 error

although you can with getInitialProps, which i'm using for another page

version: `13.4.12`

```
// middleware (from clerk)

import { authMiddleware } from '@clerk/nextjs';
import { NextResponse } from 'next/server';
import requestIp from 'request-ip';

const authOnlyRoutes = ['/library', '/trash', '/favorites'];

export default authMiddleware({
    publicRoutes: [
        '/',
        '/api/graphql',
        '/api/clerk-webhook',
        '/:handle',
        '/upload/:id',
        '/about',
        '/explore',
        '/messages',
        '/404',
        '/auth/sso-callback',
    ],

    debug: process.env.CLERK_DEBUG === 'true',
    afterAuth: (auth, req) => {
        const url = new URL(req.url);
        if (url.pathname === '/' && auth.userId) {
            return NextResponse.redirect(
                `${process.env.NEXT_PUBLIC_API_URL || ''}/explore`
            );
        }
        if (authOnlyRoutes.includes(url.pathname) && !auth.userId) {
            return NextResponse.redirect(
                `${process.env.NEXT_PUBLIC_API_URL || ''}/404`
            );
        }
    },
    beforeAuth: (request) => {
        if (request.nextUrl.pathname.startsWith('/mp')) {
            const requestHeaders = new Headers(request.headers);

            requestHeaders.set(
                'X-Forwarded-For',
                // @ts-ignore
                requestIp.getClientIp(request) || request.ip || ''
            );

            return NextResponse.rewrite(request.nextUrl, {
                request: {
                    headers: requestHeaders,
                },
            });
        }
    },
});

// Stop Middleware running on static files
export const config = {
    matcher: [
        '/((?!.*\\..*|_next).*)',
        '/',
        '/(api|graphql)(.*)',
        '/mp/:path*',
    ],
};
```

**Reply by icyJoseph** [reply]
Was kind of hoping that the middleware was running on `_next/data` , but it looks like the matcher is excluding that. 

I wonder if there's an issue with your GSSP? Though why would your about page fail too... 🤔 Like running GSSP once deployed is crashing for some reason.

**Reply by sgrund14** [reply]
hmmm, but what would the issue be? right now it's just this

```
export const getServerSideProps: GetServerSideProps = () => {
    return {
        props: {
            layout: 'default',
        },
    };
};
```

**Reply by icyJoseph** [reply]
Yeah, I don't know, assuming you pass `pageProps` to `Component` on `_app`, then we are almost running out of things to try.

If you put a log on that GSSP, does it log when you make requests? (On the Vercel Logs Dashboard)

**Reply by sgrund14** [reply]
hmm it doesn't log my log statement, but i am noticing this in the vercel logs

```
[Error: EMFILE: too many open files, open '/var/task/node_modules/@headlessui/react/dist/utils/owner.js'] {
  errno: -24,
  code: 'EMFILE',
  syscall: 'open',
  path: '/var/task/node_modules/@headlessui/react/dist/utils/owner.js'
}
[Error: EMFILE: too many open files, open '/var/task/node_modules/@headlessui/react/dist/utils/owner.js'] {
  errno: -24,
  code: 'EMFILE',
  syscall: 'open',
  path: '/var/task/node_modules/@headlessui/react/dist/utils/owner.js',
  page: '/about'
}
RequestId: 9da14011-df25-4131-bf43-8af4833d39cd Error: Runtime exited with error: exit status 1
Runtime.ExitError
```
which links back to this issue https://github.com/vercel/next.js/issues/52646 ... hmmmm... could be something to do with that? although wouldn't explain why it's also trying to load the `.json` for the about page, and why it's not happening for other pages

**Reply by icyJoseph** [reply]
Is that "about" JSON request an options call? 

And yeah, I guess that to serve your page on demand is opening too many files (?) And that crashes out trying to produce a response, either HTML (direct navigation) or JSON (for SPA transitions)

**Reply by sgrund14** [reply]
it's a HEAD call

hm, ok I'll try and fix that second issue. can't see why it is only happening on only specific routes though 🤔

**Reply by icyJoseph** [reply]
Right so that HEAD request is only for some middleware checks they do.

Do you have other GSSP routes? Strange indeed.

**Reply by sgrund14** [reply]
No, although I have some getInitialProps routes (for which SPA transitions are working, and direct navigation is broken)

half considering just deleting the vercel project entirely and starting fresh, if it is indeed something on disk. feels like that should be last resort tho hehe
